iquip/apps/scheduler.py

```diff
@@ -109,13 +190,53 @@ class SchedulerApp(qiwis.BaseApp):
     """App for fetching and showing the scheduled queue.
 
     Attributes:
+        proxy_id: The proxy server IP address.
+        proxy_port: The proxy server PORT number.
         schedulerFrame: The frame that shows the scheduled queue.
+        scheduleThread: The most recently executed _ScheduleThread instance.
     """
 
     def __init__(self, name: str, parent: Optional[QObject] = None):
         """Extended."""
         super().__init__(name, parent=parent)
+        self.proxy_ip = self.constants.proxy_ip  # pylint: disable=no-member
+        self.proxy_port = self.constants.proxy_port  # pylint: disable=no-member
+        self.scheduleThread: Optional[_ScheduleThread] = None
         self.schedulerFrame = SchedulerFrame()
+        self.startScheduleThread()
+
+    @pyqtSlot(bool, float, list)
+    def updateScheduleModel(
+        self,
+        isChanged: bool,
+        updatedTime: float,
+        schedule: Iterable[SubmittedExperimentInfo]
+    ):
+        """Updates schedulerFrame.scheduleModel using the given schedule.
+        
+        Args:
+            See _ScheduleThread signals section.
+        """
+        if isChanged:
+            model = self.schedulerFrame.scheduleModel
+            model.beginResetModel()
+            model.scheduleList = schedule
```

Comment:
Since the `ScheduleModel.scheduleList` is a `list` based on its first definition, `schedule` type hint should be `List`.
Or, based on its usage, it should be `Sequence` as it must allow arbitrary indexing.

Anyway, I think it is not a good idea to assign `schedule` directly to `model.scheduleList` since it is a mutable data structure and it came from outside (given by an argument), which means there could be many other _modifiers_.

Of course `SubmittedExperimendInfo` itself is also mutable, but I think that is okay since we would barely modify them. On the other hand, lists are prone to be modified and we might make mistakes easily.

My suggestion is to clear the `model.scheduleList` first and then update it using e.g., `extend()`. You can always adapt your own way to resolve issues though!

Reply:
Oh, it's a good point! Then, how about using just `copy()` or `deepcopy()`?